### ISSUE TITLE
Prisma not generating correct payload for types in models for MongoDB for 4.16.1 

### ISSUE THREAD

**Issue by mayerlench** [prisma/prisma#19933]
### Bug description

Prisma is generating the wrong types for types in models

There was a fix in 4.16.1 for this but it but seems to still be broken when the type is made nullable
https://github.com/prisma/prisma/issues/19890#issuecomment-1604519112

### How to reproduce

1. run prisma generate with the provided schema
2. use the message type from the prisma client
```
import { message } from "@prisma/client";

const a = {} as message;
a.data. <--- not the correct types
```
3. attempting to path into a.data will show composites, objects, scalars types

### Expected behavior

I would expect to see the types in "data" (name, age)

This is happening when "data" is made optional, if you switch to non-optional (remove the ?) it works as expected

### Prisma information

```
datasource db {
  provider = "mongodb"
  url      = env("DATABASE_URL")
}

generator client {
  provider = "prisma-client-js"
}

type data {
  name String
  age String
}

model message {
  id   String @id @default(auto()) @map("_id") @db.ObjectId
  data data?
}
```

```
import { message } from "@prisma/client";

const a = {} as message;

a.data.name
```

### Environment & setup

- OS:macOS
- Database:  MongoDB
- Node.js version: 18

### Prisma Version

```
prisma                  : 4.16.1
@prisma/client          : 4.16.1
Current platform        : darwin-arm64
Query Engine (Node-API) : libquery-engine b20ead4d3ab9e78ac112966e242ded703f4a052c (at ../../../../.nvm/versions/node/v16.18.1/lib/node_modules/prisma/node_modules/@prisma/engines/libquery_engine-darwin-arm64.dylib.node)
Migration Engine        : migration-engine-cli b20ead4d3ab9e78ac112966e242ded703f4a052c (at ../../../../.nvm/versions/node/v16.18.1/lib/node_modules/prisma/node_modules/@prisma/engines/migration-engine-darwin-arm64)
Format Wasm             : @prisma/prisma-fmt-wasm 4.16.0-66.b20ead4d3ab9e78ac112966e242ded703f4a052c
Default Engines Hash    : b20ead4d3ab9e78ac112966e242ded703f4a052c
Studio                  : 0.484.0
```

**Comment by takumiya081** [comment]
thanks! I made same bug [repo](https://github.com/takumiya081/prisma-mongodb-issue/blob/master/src/prisma-examples/6-find-posts.ts)

**Comment by millsp** [comment]
Hey @mayerlench @takumiya081 we have worked on a fix, would you mind giving it a try? You'll need to install `@prisma/client@4.16.2-dev.2` and `prisma@4.16.2-dev.2`.

**Comment by takumiya081** [comment]
@millsp 
I tried `@prisma/client@4.16.2-dev.2`, it works fine! thanks a lot 😄

**Comment by mayerlench** [comment]
> prisma@4.16.2-dev.2

Confirmed, tried it and it is generating the correct types 

Thanks!

**Comment by mayerlench** [comment]
> Hey @mayerlench @takumiya081 we have worked on a fix, would you mind giving it a try? You'll need to install `@prisma/client@4.16.2-dev.2` and `prisma@4.16.2-dev.2`.

Confirmed, tried it and it is generating the correct types

Thanks!

**Comment by Jolg42** [comment]
Note: fixed by https://github.com/prisma/prisma/pull/19920

**Comment by Jolg42** [comment]
:shipit: Note: This fix is now available as [`4.16.2`](https://github.com/prisma/prisma/releases/tag/4.16.2)
